File: services/api/app/scoring/need_score.py
"""Deterministic Need Score engine.

Designed as an isolated module so it can later be replaced with an ML model.

Need Score (0-100) =
    40% supply shortage
  + 20% estimated demand
  + 15% time since verification
  + 15% recent requests
  + 10% historical demand
"""
from dataclasses import dataclass
from datetime import datetime, timezone

WEIGHTS = {
    "supply_shortage": 0.40,
    "estimated_demand": 0.20,
    "staleness": 0.15,
    "recent_requests": 0.15,
    "historical_demand": 0.10,
}

# Normalisation caps
DEMAND_CAP = 30          # pads/day considered "very high demand"
STALENESS_CAP_HOURS = 48  # unverified for 48h => fully stale
RECENT_REQUESTS_CAP = 10  # 10+ requests in 24h => maximum pressure


@dataclass(frozen=True)
class NeedInputs:
    current_supply: int
    estimated_demand: int
    last_verified_at: datetime | None
    recent_requests_24h: int
    historical_daily_demand: float
    now: datetime | None = None


def _clamp(value: float, low: float = 0.0, high: float = 100.0) -> float:
    return max(low, min(high, value))


def supply_shortage_component(current_supply: int, estimated_demand: int) -> float:
    demand = max(estimated_demand, 1)
    ratio = current_supply / demand
    return _clamp((1 - ratio) * 100)


def demand_component(estimated_demand: int) -> float:
    return _clamp(estimated_demand / DEMAND_CAP * 100)


def staleness_component(last_verified_at: datetime | None, now: datetime) -> float:
    if last_verified_at is None:
        return 100.0
    verified = last_verified_at
    if verified.tzinfo is None:
        verified = verified.replace(tzinfo=timezone.utc)
    hours = (now - verified).total_seconds() / 3600
    return _clamp(hours / STALENESS_CAP_HOURS * 100)


def recent_requests_component(recent_requests_24h: int) -> float:
    return _clamp(recent_requests_24h / RECENT_REQUESTS_CAP * 100)


def historical_component(historical_daily_demand: float) -> float:
    return _clamp(historical_daily_demand / DEMAND_CAP * 100)


def compute_need_score(inputs: NeedInputs) -> int:
    now = inputs.now or datetime.now(timezone.utc)
    score = (
        WEIGHTS["supply_shortage"] * supply_shortage_component(inputs.current_supply, inputs.estimated_demand)
        + WEIGHTS["estimated_demand"] * demand_component(inputs.estimated_demand)
        + WEIGHTS["staleness"] * staleness_component(inputs.last_verified_at, now)
        + WEIGHTS["recent_requests"] * recent_requests_component(inputs.recent_requests_24h)
        + WEIGHTS["historical_demand"] * historical_component(inputs.historical_daily_demand)
    )
    return int(round(_clamp(score)))


def classify_need(score: int) -> str:
    if score <= 30:
        return "LOW"
    if score <= 60:
        return "MODERATE"
    if score <= 80:
        return "HIGH"
    return "CRITICAL"

File: apps/web/server/needScore.ts
// Deterministic Need Score engine — direct port of services/api/app/scoring/need_score.py
//
// Need Score (0-100) =
//     40% supply shortage
//   + 20% estimated demand
//   + 15% time since verification
//   + 15% recent requests
//   + 10% historical demand

export const WEIGHTS = {
  supply_shortage: 0.4,
  estimated_demand: 0.2,
  staleness: 0.15,
  recent_requests: 0.15,
  historical_demand: 0.1,
};

const DEMAND_CAP = 30; // pads/day considered "very high demand"
const STALENESS_CAP_HOURS = 48; // unverified for 48h => fully stale
const RECENT_REQUESTS_CAP = 10; // 10+ requests in 24h => maximum pressure

export interface NeedInputs {
  current_supply: number;
  estimated_demand: number;
  last_verified_at: Date | null;
  recent_requests_24h: number;
  historical_daily_demand: number;
  now?: Date;
}

function clamp(value: number, low = 0, high = 100): number {
  return Math.max(low, Math.min(high, value));
}

export function supplyShortageComponent(currentSupply: number, estimatedDemand: number): number {
  const demand = Math.max(estimatedDemand, 1);
  return clamp((1 - currentSupply / demand) * 100);
}

export function demandComponent(estimatedDemand: number): number {
  return clamp((estimatedDemand / DEMAND_CAP) * 100);
}

export function stalenessComponent(lastVerifiedAt: Date | null, now: Date): number {
  if (lastVerifiedAt === null) return 100;
  const hours = (now.getTime() - lastVerifiedAt.getTime()) / 3_600_000;
  return clamp((hours / STALENESS_CAP_HOURS) * 100);
}

export function recentRequestsComponent(recentRequests24h: number): number {
  return clamp((recentRequests24h / RECENT_REQUESTS_CAP) * 100);
}

export function historicalComponent(historicalDailyDemand: number): number {
  return clamp((historicalDailyDemand / DEMAND_CAP) * 100);
}

export function computeNeedScore(inputs: NeedInputs): number {
  const now = inputs.now ?? new Date();
  const score =
    WEIGHTS.supply_shortage *
      supplyShortageComponent(inputs.current_supply, inputs.estimated_demand) +
    WEIGHTS.estimated_demand * demandComponent(inputs.estimated_demand) +
    WEIGHTS.staleness * stalenessComponent(inputs.last_verified_at, now) +
    WEIGHTS.recent_requests * recentRequestsComponent(inputs.recent_requests_24h) +
    WEIGHTS.historical_demand * historicalComponent(inputs.historical_daily_demand);
  return Math.round(clamp(score));
}

export function classifyNeed(score: number): "LOW" | "MODERATE" | "HIGH" | "CRITICAL" {
  if (score <= 30) return "LOW";
  if (score <= 60) return "MODERATE";
  if (score <= 80) return "HIGH";
  return "CRITICAL";
}
